Project-Aurora/Project-Aurora/Devices/Ducky/DuckyDevice.cs

```diff
@@ -0,0 +1,360 @@
+        }
+
+        public bool IsKeyboardConnected()
+        {
+            return isInitialized;
+        }
+
+        public bool IsPeripheralConnected()
+        {
+            return isInitialized;
+        }
+
+        public bool UpdateDevice(Dictionary<DeviceKeys, Color> keyColors, DoWorkEventArgs e, bool forced = false)
+        {
+            foreach (KeyValuePair<DeviceKeys, Color> kc in keyColors)
+            {
+                //This keyboard doesn't take alpha (transparency) values, so we do this:
+                processedColor = ColorUtils.CorrectWithAlpha(kc.Value);
+
+                //This if statement grabs the packet offset from the key that Aurora wants to set, using packetOffsetMap.
+                //It also checks whether the key exists in the Dictionary, and if not, doesn't try and set the key colour.
+                if(!packetOffsetMap.TryGetValue(kc.Key, out currentOffset)){
+                    continue;
+                }
+                //The colours are encoded using RGB bytes consecutively throughout the 10 packets, which are offset with packetOffsetMap.
+                colourMessage[Packet(currentOffset[PACKET_NUM]) + currentOffset[OFFSET_NUM] + 1] = processedColor.R;
+                //To account for the headers in the next packet, the offset is pushed a further four bytes (only required if the R byte starts on the last byte of a packet).
+                if (currentOffset[OFFSET_NUM] == 63)
+                {
+                    colourMessage[Packet(currentOffset[PACKET_NUM]) + currentOffset[OFFSET_NUM] + 6] = processedColor.G;
+                    colourMessage[Packet(currentOffset[PACKET_NUM]) + currentOffset[OFFSET_NUM] + 7] = processedColor.B;
+                }
+                else
+                {
+                    colourMessage[Packet(currentOffset[PACKET_NUM]) + currentOffset[OFFSET_NUM] + 2] = processedColor.G;
+                    colourMessage[Packet(currentOffset[PACKET_NUM]) + currentOffset[OFFSET_NUM] + 3] = processedColor.B;
+                }
+            }
+
+            //Everything previous to setting the colours actually just write the colour data to the ColourMessage byte array.
+            /*
+             The keyboard is only set up to change all key colours at once, using 10 USB HID packets. They consist of:
+             One initializing packet
+             Eight colour packets (although the eighth one isn't used at all)
+             and one terminate packet
+             
+             These packets are 64 bytes each (technically 65 but the first byte is just padding, which is why there's the .Take(65) there)
+             Each key has its own three bytes for r,g,b somewhere in the 8 colour packets. These positions are defined in the packetOffsetMap
+             The colour packets also have a header. (You might be able to send these packets out of order, and the headers will tell the keyboard where it should be, but IDK)*/
+            for (int i = 0; i < 10; i++)
```

Comment:
You could store the last array you sent and only send it again if the colors changed, maybe that helps with the keyboard not responding after a while.